src/Symfony/Component/Security/Http/RememberMe/RememberMeDetails.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\RememberMe;
+
+use Symfony\Component\Security\Core\Authentication\RememberMe\PersistentToken;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class RememberMeDetails
+{
+    public const COOKIE_DELIMITER = ':';
+
+    private $userFqcn;
+    private $username;
+    private $expires;
+    private $value;
+
+    public function __construct(string $userFqcn, string $username, int $expires, string $value)
+    {
+        $this->userFqcn = $userFqcn;
+        $this->username = $username;
+        $this->expires = $expires;
+        $this->value = $value;
+    }
+
+    public static function fromRawCookie(string $rawCookie): self
+    {
+        $cookieParts = explode(self::COOKIE_DELIMITER, base64_decode($rawCookie), 4);
+        if (false === $cookieParts[1] = base64_decode($cookieParts[1], true)) {
+            throw new AuthenticationException('Username contains a character from outside the base64 alphabet.');
```

Comment:
`BadCredentialsException`?

Reply:
I think this is an impossible code path (and purely added as the current remember me system also does this check and strictly, `base64_encode()` can return `false`). As there is nothing a user can do about this, I don't think a `BadCredentialsException` makes sense here (but I'm happy to hear other ways `BadCredentialsException` could be useful).

For reference, it's `AuthenticationException` in the 5.2 code: https://github.com/symfony/symfony/blob/f052d11a43f00c9d0233033954657743fca8df0a/src/Symfony/Component/Security/Http/RememberMe/TokenBasedRememberMeServices.php#L39-L41